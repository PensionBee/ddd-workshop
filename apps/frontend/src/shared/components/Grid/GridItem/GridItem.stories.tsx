import React from "react";
import { ThemeProvider } from "styled-components";
import type { ComponentMeta, ComponentStory } from "@storybook/react";
import { useMediaQueryMd } from "@shared/hooks/useMediaQuery/useMediaQuery";
import { GlobalStyles } from "@shared/styling/globalStyles";
import { theme } from "@shared/styling/theme";
import { Box } from "../../Box/Box";
import { GridContainer } from "../GridContainer/GridContainer";
import { GridRow } from "../GridRow/GridRow";
import type { GridItemProps } from "./GridItem";
import { GridItem } from "./GridItem";
import { CaptionL } from "../../Typography/Typography";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const themeDecorator = (Story: any) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      <Box>
        <Story />
      </Box>
    </ThemeProvider>
  );
};

export default {
  title: "Shared/Grid/Item",
  component: GridItem,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
  decorators: [themeDecorator],
} as ComponentMeta<typeof GridItem>;

const GridItemTemplate: ComponentStory<typeof GridItem> = ({
  children,
  ...args
}: GridItemProps) => {
  return <GridItem {...args}>{children}</GridItem>;
};

const Child = ({ h }: { h?: string }) => (
  <Box backgroundColor="grey4" w="100%" h={h || 200} />
);

const GridItemDefault = GridItemTemplate.bind({});
GridItemDefault.args = {
  children: <Child />,
};

export const OneItem = () => {
  return (
    <GridContainer backgroundColor="grey8">
      <GridRow rowGap={20}>
        <GridItem>
          <Child />
        </GridItem>
      </GridRow>
    </GridContainer>
  );
};

export const TwoItems = () => {
  const items = [0, 1];
  return (
    <GridContainer backgroundColor="grey8">
      <GridRow rowGap={20}>
        {items.map((item) => (
          <GridItem key={item} span={6}>
            <Child />
          </GridItem>
        ))}
      </GridRow>
    </GridContainer>
  );
};

export const ThreeItems = () => {
  const items = [0, 1, 2];
  return (
    <GridContainer backgroundColor="grey8">
      <GridRow rowGap={20}>
        {items.map((item) => (
          <GridItem key={item} span={4}>
            <Child />
          </GridItem>
        ))}
      </GridRow>
    </GridContainer>
  );
};

export const FourItems = () => {
  const items = [0, 1, 2, 3];
  return (
    <GridContainer backgroundColor="grey8">
      <GridRow rowGap={20}>
        {items.map((item) => (
          <GridItem key={item} span={3}>
            <Child />
          </GridItem>
        ))}
      </GridRow>
    </GridContainer>
  );
};

export const TwelveItems = ({ h }: { h?: string }) => {
  const items = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];

  return (
    <GridContainer backgroundColor="grey8">
      <GridRow rowGap={20}>
        {items.map((item) => (
          <GridItem key={item} span={{ xs: 3, md: 1 }}>
            <Child h={h} />
          </GridItem>
        ))}
      </GridRow>
    </GridContainer>
  );
};

const ItemContent: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <Box w="100%" justify="center" d="flex" backgroundColor="grey4">
      <CaptionL>{children}</CaptionL>
    </Box>
  );
};

export const ItemSpan = () => {
  const isNotMobile = useMediaQueryMd();

  if (isNotMobile) {
    return (
      <GridContainer>
        <GridRow pb={20}>
          <GridItem span={1}>
            <ItemContent>1</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={2}>
            <ItemContent>2</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={3}>
            <ItemContent>3</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={4}>
            <ItemContent>4</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={5}>
            <ItemContent>5</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={6}>
            <ItemContent>6</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={7}>
            <ItemContent>7</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={8}>
            <ItemContent>8</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={9}>
            <ItemContent>9</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={10}>
            <ItemContent>10</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={11}>
            <ItemContent>11</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={12}>
            <ItemContent>12</ItemContent>
          </GridItem>
        </GridRow>
      </GridContainer>
    );
  }

  return (
    <GridContainer>
      <GridRow pb={20}>
        <GridItem span={3}>
          <ItemContent>3</ItemContent>
        </GridItem>
      </GridRow>

      <GridRow pb={20}>
        <GridItem span={6}>
          <ItemContent>6</ItemContent>
        </GridItem>
      </GridRow>

      <GridRow pb={20}>
        <GridItem span={9}>
          <ItemContent>9</ItemContent>
        </GridItem>
      </GridRow>

      <GridRow pb={20}>
        <GridItem span={12}>
          <ItemContent>12</ItemContent>
        </GridItem>
      </GridRow>
    </GridContainer>
  );
};

export const ItemOffset = () => {
  const isNotMobile = useMediaQueryMd();

  if (isNotMobile) {
    return (
      <GridContainer>
        <GridRow pb={20}>
          <GridItem span={1} offset={0}>
            <ItemContent>0</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={1}>
            <ItemContent>1</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={2}>
            <ItemContent>2</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={3}>
            <ItemContent>3</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={4}>
            <ItemContent>4</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={5}>
            <ItemContent>5</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={6}>
            <ItemContent>6</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={7}>
            <ItemContent>7</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={8}>
            <ItemContent>8</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={9}>
            <ItemContent>9</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={10}>
            <ItemContent>10</ItemContent>
          </GridItem>
        </GridRow>

        <GridRow pb={20}>
          <GridItem span={1} offset={11}>
            <ItemContent>11</ItemContent>
          </GridItem>
        </GridRow>
      </GridContainer>
    );
  }

  return (
    <GridContainer>
      <GridRow pb={20}>
        <GridItem span={3} offset={0}>
          <ItemContent>0</ItemContent>
        </GridItem>
      </GridRow>

      <GridRow pb={20}>
        <GridItem span={3} offset={3}>
          <ItemContent>3</ItemContent>
        </GridItem>
      </GridRow>

      <GridRow pb={20}>
        <GridItem span={3} offset={6}>
          <ItemContent>6</ItemContent>
        </GridItem>
      </GridRow>

      <GridRow pb={20}>
        <GridItem span={3} offset={9}>
          <ItemContent>9</ItemContent>
        </GridItem>
      </GridRow>
    </GridContainer>
  );
};
