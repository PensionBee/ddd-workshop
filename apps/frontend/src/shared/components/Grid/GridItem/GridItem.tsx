import type {
  BreakpointValues,
  ValueOrBreakpointValues,
} from "@shared/utils/types";
import { Box } from "../../Box/Box";
import { asBreakpointObject } from "@shared/utils/asBreakpointObject";
import type { GridItemOffset, GridItemSpan } from "./GridItem.types";
import type {
  MediaDirectionType,
  MediaFlexType,
  MediaMarginType,
  MediaWidthType,
} from "@shared/styling/styledComponentHelpers";

export type GridItemProps = {
  /** React children */
  children: React.ReactNode;
  /** Element to render as */
  as?: React.ComponentProps<typeof Box>["as"];
  /** Horizontal alignment */
  align?: React.ComponentProps<typeof Box>["align"];
  /** Column offset */
  offset?: ValueOrBreakpointValues<GridItemOffset>;
  /** Column span */
  span?: ValueOrBreakpointValues<GridItemSpan>;
  /** Background color */
  backgroundColor?: React.ComponentProps<typeof Box>["backgroundColor"];
  /** Flex direction */
  direction?: ValueOrBreakpointValues<MediaDirectionType>;
};

const defaultOffset: ValueOrBreakpointValues<GridItemOffset> = { xs: 0 };
const defaultSpan: ValueOrBreakpointValues<GridItemSpan> = { xs: 12 };

export const GridItem: React.FC<GridItemProps> = ({
  children,
  as = "div",
  align = "stretch",
  offset = defaultOffset,
  span = defaultSpan,
  backgroundColor,
  direction = "column",
  ...props
}) => {
  const transformedProps = {
    ...transformOffsetToBoxStyles({
      ...defaultOffset,
      ...asBreakpointObject(offset),
    }),
    ...transformSpanToBoxStyles({
      ...defaultSpan,
      ...asBreakpointObject(span),
    }),
  };

  const styledProps = {
    $direction: asBreakpointObject(direction),
  };

  return (
    <Box
      data-rc="GridItem"
      align={align}
      as={as}
      d="flex"
      direction="column"
      px={{ xs: 12, md: 14, lg: 24 }}
      backgroundColor={backgroundColor}
      {...transformedProps}
      {...styledProps}
      {...props}
    >
      {children}
    </Box>
  );
};

export const transformOffsetToBoxStyles = (
  offset: BreakpointValues<GridItemOffset>
) =>
  Object.entries(offset).reduce(
    (prev, [breakpoint, colsToOffset]) => {
      const ml = {
        ...prev.ml,
        [breakpoint]: (colsToOffset / 12) * 100 + "%",
      };
      return { ...prev, ml };
    },
    { ml: {} satisfies BreakpointValues<MediaMarginType> }
  );

export const transformSpanToBoxStyles = (
  span: BreakpointValues<GridItemSpan>
) =>
  Object.entries(span).reduce(
    (prev, [breakpoint, colsToSpan]) => {
      const percentageWidth = `${(colsToSpan / 12) * 100}%`;
      const flex = {
        ...prev.flex,
        [breakpoint]: `0 0 ${percentageWidth}`,
      };
      const wMax = {
        ...prev.wMax,
        [breakpoint]: percentageWidth,
      };
      return { ...prev, flex, wMax };
    },
    {
      flex: {} satisfies BreakpointValues<MediaFlexType>,
      wMax: {} satisfies BreakpointValues<MediaWidthType>,
    }
  );
