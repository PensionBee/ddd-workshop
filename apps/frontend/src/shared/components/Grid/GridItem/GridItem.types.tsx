export type GridItemOffset = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
export type GridItemSpan = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
