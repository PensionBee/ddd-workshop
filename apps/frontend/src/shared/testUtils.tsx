import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
