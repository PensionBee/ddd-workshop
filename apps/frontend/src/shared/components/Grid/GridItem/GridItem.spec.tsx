import { render } from "@shared/testUtils";
import { GridItem } from "./GridItem";

describe("GridItem", () => {
  const colContent = "GridItem Content";
  describe("Default", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <GridItem>{colContent}</GridItem>
      );

      expect(getByText(colContent)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render as elements", () => {
      const { container, asFragment } = render(
        <GridItem as="span">{colContent}</GridItem>
      );

      expect(container.getElementsByTagName("span").length).toBe(1);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
