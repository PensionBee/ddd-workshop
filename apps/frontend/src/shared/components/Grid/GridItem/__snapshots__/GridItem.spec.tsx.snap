// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridItem Default should render as elements 1`] = `
<DocumentFragment>
  <span
    class="sc-bcXHqe hYpwdP"
    data-rc="GridItem"
  >
    GridItem Content
  </span>
</DocumentFragment>
`;

exports[`GridItem Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe hYpwdP"
    data-rc="GridItem"
  >
    GridItem Content
  </div>
</DocumentFragment>
`;
